Fix two problems with `unset -f` behavior

JohnoKing opened the pull request:
This pull request backports two bug fixes from ksh93v- for the problems detailed in #21. These fixes do the following:

In `src/cmd/ksh93/sh/name.c`, the check for if a function is still running has been corrected to fix a segmentation fault that occurred when a POSIX function unset itself.

The missing functionality for when a KornShell style function unsets itself has been added to `src/cmd/ksh93/sh/xec.c`. This kind of function will now be properly unset after it has finished running, rather than silently cause `unset -f` to fail.

McDutchie commented on NEWS:
```diff
@@ -3,6 +3,14 @@ For full details, see the git log at: https://github.com/ksh93/ksh
 
 Any uppercase BUG_* names are modernish shell bug IDs.
 
+2020-06-17:
```
Please add a blank line for formatting consistency.

JohnoKing replied:
This has been fixed.

McDutchie commented on src/cmd/ksh93/tests/functions.sh:
```diff
@@ -1249,5 +1249,10 @@ expect_status=0
+# ======
+# When a function unsets itself, it should not fail to be unset
+$SHELL -c 'fn() { unset -f fn; true; }; fn' 2> /dev/null || err_exit 'unset of POSIX function in the calling stack fails'
+$SHELL -c 'function ftest { ftest2; }; function ftest2 { unset -f ftest; }; ftest' 2> /dev/null || err_exit 'unset of ksh function in the calling stack fails'
```
These tests are not complete. They check that the crash is gone, but they don't actually check whether unsetting the function succeeded.

There are various ways of checking that, but the most reliable is probably the brute approach: set `PATH=/dev/null`, then try to run the function a second time with `2>/dev/null` and check that the exit status is 127 (not found).

I'd also lose the global `2>/dev/null` on these tests. If `unset -f` ever prints an error message, I would want to know about that.

JohnoKing replied:
I have made the requested changes to the two regression tests. I have also added another test as well for another potential bug in POSIX functions (although the new test does use `2> /dev/null` as it is expected to return an error).

JohnoKing replied:
Oops, I forgot to check the return status. One moment.

JohnoKing replied:
The regression tests in this pull request (there are now four added) have been fully updated.

McDutchie reviewed (changes requested):
I was half expecting you to show up for this one. :)  Couple of comments below…

JohnoKing committed:
Fix two problems with `unset -f` behavior

src/cmd/ksh93/sh/name.c:
 - Correct the check for when a function is currently running
   to fix a segmentation fault that occurred when a POSIX
   function tries to unset itself while it is running.
   This bug fix was backported from ksh93v-.

src/cmd/ksh93/sh/xec.c:
 - If a function tries to unset itself, unset the function
   with `_nv_unset(np, NV_RDONLY)` to fix a silent failure.
   This fix was also backported from ksh93v-.

src/cmd/ksh93/tests/functions.sh:
 - Add four regression tests for when a function unsets itself.
